### PR TITLE
Avoid overshooting watermarks during relocation.

### PR DESCRIPTION
**Pull request by kovrus**
## Checklist

 - [ ] User relevant changes are recorded in ``CHANGES.txt``
 - [ ] Touched code is covered by tests
 - [ ] Documentation has been updated if necessary
 - [ ] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [ ] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by kovrus** [comment]
shall we backport it to `4.0`?

**Comment by mfussenegger** [comment]
> shall we backport it to 4.0?

yep I think we should backport it.

File: es/es-server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java [hunk 1]
```diff
@@ -272,21 +272,27 @@ private void maybeRefresh() {
         }
     }
 
+    // allow tests to adjust the node stats on receipt
+    List<NodeStats> adjustNodesStats(List<NodeStats> nodeStats) {
```
Comment (mfussenegger) [inline es/es-server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java]: I think I don't understand what's going on here. This method is only returning the argument back without doing anything?

File: sql/src/test/java/io/crate/integrationtests/DiskUsagesITest.java [hunk 2]
```diff
@@ -293,14 +294,18 @@ public void testMovesShardsOffSpecificDataPathAboveWatermark() throws Exception
 
         // there should be one shard on a bad path on node0
         execute(
-            "SELECT count(*) " +
+            "SELECT path " +
             "FROM sys.shards " +
             "WHERE table_name = 't' " +
             "  AND schema_name = 'doc'" +
-            "  AND node['id'] = ?" +
-            "  AND path LIKE concat(?, '%')",
```
Comment (kovrus) [inline sql/src/test/java/io/crate/integrationtests/DiskUsagesITest.java]: i did it with `like` originally, but it seems that it produces wrong regex patters on windows

File: es/es-server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderTests.java [hunk 3]
```diff
@@ -0,0 +1,973 @@
...
+import static org.hamcrest.Matchers.nullValue;
+
+public class DiskThresholdDeciderTests extends ESAllocationTestCase {
+
+    private DiskThresholdDecider makeDecider(Settings settings) {
+        return new DiskThresholdDecider(settings, new ClusterSettings(settings, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS));
+    }
+
+    @Test
+    public void testDiskThreshold() {
```
[the remark this answers is not in the document]
Comment (kovrus) [reply es/es-server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/DiskThresholdDeciderTests.java]: as discussed, i'll keep it as it is for now. it is a bit hard to split and cumbersome to split it.

**Pull request merged**